### PR TITLE
DoRA fixes

### LINKED ISSUE #2129: Distributed DoRA training is broken
Reported by @SLR722. Repro:

```
tune run --nnodes 1 --nproc_per_node 2 lora_finetune_distributed --config llama3/8B_dora
...
[rank1]:   File "/data/users/ebs/ebs-torchtune-alt/recipes/lora_finetune_distributed.py", line 929, in recipe_main
[rank1]:     recipe.setup(cfg=cfg)
[rank1]:   File "/data/users/ebs/ebs-torchtune-alt/recipes/lora_finetune_distributed.py", line 269, in setup
[rank1]:     self._model = self._setup_model(
[rank1]:                   ^^^^^^^^^^^^^^^^^^
[rank1]:   File "/data/users/ebs/ebs-torchtune-alt/recipes/lora_finetune_distributed.py", line 527, in _setup_model
[rank1]:     training.validate_no_params_on_meta_device(model)
[rank1]:   File "/data/users/ebs/ebs-torchtune-alt/torchtune/training/_distributed.py", line 180, in validate_no_params_on_meta_device
[rank1]:     raise RuntimeError(f"Unexpected param or buffer {n} on meta device.")
```

I've confirmed that this was broken by #1909 (fix one bug and introduce another..). That PR fixed a bug where we weren't properly initializing the magnitude in the single device recipe. Unfortunately now the magnitude is left on meta device (I think because we are now doing a copy instead of creating a new magnitude parameter).

A prerequisite for closing this issue is adding a test for DoRA to [test_lora_finetune_distributed.py](https://github.com/pytorch/torchtune/blob/26b2200010a37474015925c5e3f4606435b72dd3/tests/recipes/test_lora_finetune_distributed.py#L4)

### PR DESCRIPTION
**Pull request by ebsmothers**
Co-authored by: @mirceamironenco 

Fixes #2129 

This PR fixes our broken DoRA distributed training. It also makes some quality-of-life improvements to how we initialize `DoRALinear` and (in a small way) `LoRALinear`.

We do the following:

1) Define a `to_empty` method on both `DoRALinear` and `LoRALinear`. This consolidates the move of all the potentially-still-on-meta-device params onto a real device in a single place (`lora_a.weight` and `lora_b.weight` for LoRA, and additionally `magnitude` for DoRA) and matches the [contract of nn.Module](https://github.com/pytorch/pytorch/blob/main/torch/nn/modules/module.py#L1181). For `DoRALinear`'s magnitude we create a new parameter with `torch.empty_like` and manually set `requires_grad`, then use the `torch.utils.swap_tensors` API to ensure `self.magnitude` winds up on a real device.

2) For DoRA, we modify the `initialize_dora_magnitude` method to (a) check that base weight and LoRA weights are not on meta device, (b) calculate ||W + (alpha/rank)BA||, and (c) copy this value into `self.magnitude`. Since this should always be called after `to_empty`, we now know that `self.magnitude` 

**BC note:** From the perspective of someone importing from our library to their own custom recipe, the changes to `LoRALinear`'s contract with respect to `to_empty` are backwards compatible: calling `lora_linear.to_empty(device)` (new version) vs `lora_linear.lora_a.to_empty(device); lora_linear.lora_b.to_empty(device)` (old version) are identical. For `DoRALinear`, the recipe will be broken without the update to the new version.. but the recipe was broken anyways, so I claim this is OK.

Other changes:
- Deprecate (but don't yet delete) `load_dora_magnitudes`, which is redundant.
- Add unit tests for proper calculation of DoRA magnitude on single-device and distributed w/ meta device init

### Unit tests

```
pytest tests/torchtune/modules/peft/test_dora.py
...
======== 12 passed in 11.73s ========
```

#### Single-device recipe

Compare the following three cases:

1) This PR
2) Resume intermediate checkpoint on this PR
3) Baseline (run on main)

Loss curves for all three are the same:

Screenshot 2024-12-11 at 11 22 24 AM

Command for (1):

```
tune run lora_finetune_single_device --config llama3/8B_dora_single_device max_steps_per_epoch=100 epochs=2 \
metric_logger=torchtune.training.metric_logging.WandBLogger metric_logger.project=test-dora-fixes \
metric_logger.name=dora-single-new
```

Commands for (2):

```
# First delete final checkpoint to automatically resume from intermediate checkpoint
rm -r /tmp/torchtune/llama3_8B/dora_single_device/epoch_1

tune run lora_finetune_single_device --config llama3/8B_dora_single_device max_steps_per_epoch=100 epochs=2 \
metric_logger=torchtune.training.metric_logging.WandBLogger metric_logger.project=test-dora-fixes \
metric_logger.name=dora-single-new-resumed resume_from_checkpoint=True
```

#### Distributed recipe

1) This PR
2) Resume intermediate checkpoint on this PR

(Note that (3) from the single-device recipe is not applicable, as this is broken on main). Loss curves are the same after resuming from checkpoint:

Screenshot 2024-12-11 at 11 25 25 AM

Command for (1):

```
tune run --nproc_per_node 2 lora_finetune_distributed --config llama3/8B_dora max_steps_per_epoch=100 epochs=2 \
metric_logger=torchtune.training.metric_logging.WandBLogger metric_logger.project=test-dora-fixes \
metric_logger.name=dora-distributed-new
```

Commands for (2):

```
# First delete final checkpoint to automatically resume from intermediate checkpoint
rm -r /tmp/torchtune/llama3_8B/dora/epoch_1

tune run --nproc_per_node 2 lora_finetune_distributed --config llama3/8B_dora max_steps_per_epoch=100 epochs=2 \
metric_logger=torchtune.training.metric_logging.WandBLogger metric_logger.project=test-dora-fixes \
metric_logger.name=dora-distributed-new-resume resume_from_checkpoint=True
```

### CODE REVIEW

File: tests/torchtune/modules/peft/test_dora.py [hunk 1]
```diff
@@ -221,3 +230,176 @@ def test_quantized_state_dict(self, dtype):
...
+        if is_rank_zero:
+            for dora_linear in [ffn.w1, ffn.w2, ffn.w3]:
+                assert dora_linear.weight.is_meta
+                assert dora_linear.lora_a.weight.is_meta
+                assert dora_linear.lora_b.weight.is_meta
+                assert dora_linear.magnitude.is_meta
+
+        # Optionally load adapter weights (as though we are resuming from checkpoint)
+        # Now lora_a, lora_b, and magnitude should not be on meta device, but base weight should be
+        # Additionally since the weights are randomly initialized we should have magnitude != ||W+(alpha/rank)BA||
```
Comment (ebsmothers) [inline tests/torchtune/modules/peft/test_dora.py:331]: todo: clean up this comment

File: torchtune/modules/peft/dora.py [hunk 2]
```diff
@@ -104,11 +104,28 @@ def initialize_dora_magnitude(self):
...
+        This must be called after loading/initializing base model and LoRA params.
         """
+        if any(
+            [
+                self.weight.is_meta,
+                self.lora_a.weight.is_meta,
+                self.lora_b.weight.is_meta,
+            ]
+        ):
+            raise ValueError(
```
Comment (SalmanMohammadi) [inline torchtune/modules/peft/dora.py:117]: cc @joecummings on our docstring linter being broken

Could you add a `Raise` in the docstring here please? @ebsmothers

File: torchtune/modules/peft/dora.py [hunk 3]
```diff
@@ -104,11 +104,28 @@ def initialize_dora_magnitude(self):
...
+                "Cannot initialize DoRA magnitude until after base and LoRA parameters"
+            )
         base_weight = self.weight.to(self.lora_a.weight.dtype)
         lora_weight = self.lora_b.weight @ self.lora_a.weight
-        weight_norm = self._get_weight_norm(base_weight, lora_weight)
-        self.magnitude.copy_(weight_norm)
+
+        magnitude = nn.Parameter(
+            torch.empty_like(self.magnitude, device=self.lora_a.weight.device),
+            requires_grad=self.magnitude.requires_grad,
```
Comment (SalmanMohammadi) [inline torchtune/modules/peft/dora.py:125]: When would this be `False`?

File: torchtune/modules/peft/dora.py [hunk 4]
```diff
@@ -104,11 +104,28 @@ def initialize_dora_magnitude(self):
...
+        )
+        magnitude = self._get_weight_norm(base_weight, lora_weight)
+        torch.utils.swap_tensors(self.magnitude, magnitude)
```
Comment (SalmanMohammadi) [inline torchtune/modules/peft/dora.py:128]: not 100% sure here, but will the original `torch.empty(out_dim)` still be floating around after this? This probably isn't an issue?

File: torchtune/modules/peft/dora.py [hunk 5]
```diff
@@ -117,8 +134,7 @@ def _get_weight_norm(self, weight, lora_weight):
 
     def adapter_params(self) -> List[str]:
         """
-        Return lora_a.weight and lora_b.weight as adapter params.
-        If bias is enabled, also return lora_a.bias and lora_b.bias.
+        Return lora_a.weight, lora_b.weight, and magnitude as adapter params.
```
Comment (SalmanMohammadi) [inline torchtune/modules/peft/dora.py:137]: Could you add a proper docstring here, please?

File: tests/torchtune/modules/peft/test_dora.py [hunk 6]
```diff
@@ -221,3 +230,175 @@ def test_quantized_state_dict(self, dtype):
...
+            * dora_linear.lora_b.weight
+            @ dora_linear.lora_a.weight,
+            dim=1,
+        )
+        assert not torch.allclose(dora_linear.magnitude, expected_magnitude)
+        dora_linear.initialize_dora_magnitude()
+        assert torch.allclose(dora_linear.magnitude, expected_magnitude)
+
+
+class TestDistributedDoRALinear(FSDPTest):
```
Comment (SalmanMohammadi) [inline tests/torchtune/modules/peft/test_dora.py:259]: This is very nice

Is it worth adding a test to make sure the `ValueError`occurs when we try to initialize dora magnitudes on meta device?

File: torchtune/modules/peft/_utils.py [hunk 7]
```diff
@@ -311,14 +311,3 @@ def validate_missing_and_unexpected_for_lora(
                 raise AssertionError(f"Missing LoRA key {k} from adapter state dict")
     if lora_unexpected:
         raise AssertionError("Unexpected key loading adapter")
-
-
-def load_dora_magnitudes(model: nn.Module) -> None:
```
Comment (SalmanMohammadi) [inline torchtune/modules/peft/_utils.py:316]: Deprecate instead of delete?

File: tests/torchtune/modules/peft/test_dora.py (hunk 6 above)
Comment (ebsmothers) [reply tests/torchtune/modules/peft/test_dora.py:259]: Yeah I can add this

File: torchtune/modules/peft/dora.py (hunk 4 above)
Comment (ebsmothers) [reply torchtune/modules/peft/dora.py:128]: Yeah I was wondering about this too, need to check myself. Offhand not sure about the implications for refcounting with this utility. But magnitude is just embed_dim number of elements so even if still hanging around it’s pretty small

File: torchtune/modules/peft/dora.py (hunk 3 above)
Comment (ebsmothers) [reply torchtune/modules/peft/dora.py:125]: Offhand I can’t think of a case but imo better to just be explicit here

**Review by joecummings — APPROVED** [review]
